apps/sms/views/shared/js/navigation.js

```diff
@@ -1,361 +1,921 @@
+    debug('Nothing to execute for step', stepName);
     return Promise.resolve();
-  },
+  }
 
-  /**
-   * Lifecycle methods are called in this order:
-   * - previousPanel.beforeLeave
-   * - nextPanel.beforeEnter
-   * - previousPanel.afterLeave
-   * - nextPanel.afterEnter
-   *
-   * The panels are slid between beforeEnter and afterLeave.
-   *
-   * Therefore it is expected that:
-   * - beforeLeave does not change any state but merely checks whether the
-   *   transition can take place
-   * - beforeEnter checks whether the transition can take place and make any
-   *   visual change that should be made before the transition
-   * - afterLeave cleans up the panel, reclaim memory
-   * - afterEnter does left-over actions
-   *
-   * beforeLeave and beforeEnter can :
-   * - throw or return a rejected Promise to stop the transition.
-   * - finish normally or return a resolved Promise to continue the transition
-   *
-   * However, afterLeave and afterEnter should only finish normally or return a
-   * resolved Promise. Otherwise, the errors will be logged but the transition
-   * will finish.
-   *
-   * toPanel itself returns a Promise that will be resolved if the transition
-   * happened and rejected otherwise.
-   */
-  toPanel: function n_toPanel(panel, args) {
-    if (!(panel in this.panels)) {
-      return Promise.reject(new Error('Panel ' + panel + ' is unknown.'));
-    }
+  var panel = state.panel;
 
-    // put the request to a queue when:
-    //   - Panel is still transitioning
-    //   - trying to navigate when the app is not loaded completely
-    var notReadyNavigate = (panel !== this.defaultPanel && !this.isReady);
-    if (this.transitionPromise || notReadyNavigate) {
-      queuedPanel = {
-        panel: panel,
-        args: args
-      };
-
-      return new Promise(function(resolve, reject) {
-        queuedPanel.defer = {
-          resolve: resolve,
-          reject: reject
-        };
-      });
-    }
+  var resultPromise = Promise.resolve();
+  var viewObject = window[panel.behavior];
+  if (viewObject[stepName]) {
+    // encapsulating as a then handler to catch exceptions.
+    resultPromise = Promise.resolve().then(
+      () => viewObject[stepName](
+        navigationTransition.transitionArgs
+      )
+    );
+  }
 
-    // this checks both panel and args
-    if (this.isCurrentPanel(panel, args)) {
-      return Promise.resolve();
-    }
+  return resultPromise.then(
+    () => navigationTransition.stepsRun.push(stepName)
+  ).then(
+    () => debug(`Finished executeNavigationStep(${stepName})`)
+  );
+}
 
-    var transitionArgs = args || {};
+/**
+ * Attach a handler to run the `afterEnter` step after the page is loaded.
+ * It uses either the custom event `navigation-transition-end` or the usual
+ * `load` event.
+ * @returns {Promise} Resolved once afterEnter is run.
+ */
+function attachAfterEnterHandler() {
+  if (document.readyState === 'complete') {
+    return Promise.resolve().then(
```

Comment:
I think I added this when executeNavigationStep was not catching exceptions yet -- problems of getting interrupted in the middle of a change :) Thanks !